# Movies Tonight Part 3: SQL DML

In [8]:
%load_ext sql
%sql sqlite:///Movies.db

The sql extension is already loaded. To reload it, use:
  %reload_ext sql


'Connected: @Movies.db'

1. **Load the `DATASET.csv` file into the `Movies.db` database.**  
    - Import the CSV file to create a new table called `DATASET`. Follow the example [here](http://www.sqlitetutorial.net/sqlite-import-csv/).
    - Use a [`PRAGMA table-info` pseudo-query](https://www.sqlite.org/pragma.html#pragma_table_info) to inspect the design of the new table. [How are we going to handle dates](https://www.sqlite.org/datatype3.html)?

2. **Populate the tables with data.**  

    - Use `INSERT ... SELECT ...` queries to extract data from `DATASET` for each table of your other tables. 
    - Start with the strong entities before moving on to the weak ones. Be sure to always select distinct     values. 
    - Foreign Keys get special treatment. You will have to use JOINs to pick up the PKs from the other side of the relationships. (As in Part 1, use the fact that movie names, theater names, and people names are always unique in our source data. That way you can use then for the JOIN columns. 
    - If you have to start over, then rerun your DDL queries from Part 2. 

In [22]:
%%sql
INSERT INTO ARTISTS (AName)
SELECT DISTINCT CName FROM DATA;

 * sqlite:///Movies.db
153 rows affected.


[]

In [15]:
%%sql
INSERT INTO MOVIES (MTitle, Rating)
    SELECT DISTINCT MTitle, Rating FROM DATA;

 * sqlite:///Movies.db
23 rows affected.


[]

In [26]:
%%sql 
INSERT INTO THEATRES (TName, Location, Phone)
SELECT DISTINCT TName, Location , Phone FROM DATA;

 * sqlite:///Movies.db
9 rows affected.


[]

In [36]:
%%sql
INSERT INTO CREDITS (MID,AID,CCode)
SELECT DISTINCT MID,AID,CCode
FROM MOVIES
    JOIN DATA USING (MTitle)
    JOIN ARTISTS ON (DATA.CName
                = ARTISTS.AName);

 * sqlite:///Movies.db
162 rows affected.


[]

In [41]:
%%sql
INSERT INTO SHOWS (MID,TID,STime)
SELECT DISTINCT MID, TID, ShowTime
FROM DATA
    JOIN MOVIES ON (MOVIES.MTitle=DATA.MTitle)
    JOIN THEATRES ON (THEATRES.TNAME=DATA.TNAME);

 * sqlite:///Movies.db
131 rows affected.


[]

3. **Check your work.**  
There are 9 theaters, 23 movies, 152 artists, 131 shows, and 161 credits.

 * sqlite:///Movies.db
(sqlite3.OperationalError) no such column: MOVIES [SQL: 'SELECT COUNT (MOVIES) FROM DATA;'] (Background on this error at: http://sqlalche.me/e/e3q8)


4. **Now check the source data.**  
There are actually 13 duplicate rows in the original source data! Can you spot them? Write a SQL query that counts how many times each (MTitle, TName, Showtime, CCode, and CName) appears. Which ones appear more than once?

In [71]:
%%sql
SELECT DISTINCT COUNT (MTitle) FROM DATA;

 * sqlite:///Movies.db
Done.


[(913,)]

5. **Drop the `DATASET` table.**  
We don't need it anymore. 

6. **Save, commit, and push your work to GitHub.**  
Use the commit summary comment "Completed Part 3".